proposals/0010-validating-resource-container-elements.md

```diff
@@ -0,0 +1,142 @@
+* Proposal: [0010](0010-validating-resource-container-elements.md)
+* Author(s): [Joshua Batista](https://github.com/bob80905)
+* Sponsor: TBD
+* Status: **Under Consideration**
+* Impacted Project(s): (LLVM)
+
+*During the review process, add the following fields as needed:*
+
+* PRs: [#NNNN](https://github.com/microsoft/DirectXShaderCompiler/pull/NNNN)
+* Issues: [#75676](https://github.com/llvm/llvm-project/issues/75676)
+
+## Introduction
+Resources are often used in HLSL, with various resource element types (RETs).
+For example:
+```
+RWBuffer<float> rwbuf: register(u0);
+```
+In this code, the RET is `float`, and the resource type is `RWBuffer`. The
+resource type is not a `RawBuffer` variant, and so there is a distinct set
+of rules that define valid RETs for this resource type.
+
+RETs for non-`RawBuffer` variants may include basic types (ints and uints of sizes 16
+and 32, as well as half, and float). Structs that contain fields of these basic
+types (where all fields in the struct have the same type) may also be RETs. 
+Structs that either have structs as fields or arrays of structs as fields may also be
+allowed, as long as everything can fit in 4 32-bit quantities. Additionally, resource 
+types are not allowed within an RET, even if the underlying resource type has a 
+primitive RET.
+
+RETs for `RawBuffer` variants are much less constrained, the only rule is that the RET
+may not be an incomplete type (a handle type or a resource type).
+
+If someone writes `RWBuffer<MyCustomType>` and MyCustomType is not a 
+valid RET, there there should be infrastructure to reject this RET and emit a message 
+explaining why it was rejected as an RET.
+
+## Motivation
+
+Currently, in `clang\lib\CodeGen\CGHLSLRuntime.cpp`, under `calculateElementType` 
+there is a whitelist of valid RETs. 
+Anything that is not an int, uint, nor a floating-point type, will be rejected.
+The whitelist isn't broad enough, because it doesn't include the case where the RET 
+is user-defined. Ideally, a user should be able to determine how any user-defined 
+structure is invalid as an RET. Some system should be in place to more completely 
+enforce the rules for valid and invalid RETs, as well as provide useful information
+on why they are invalid.
+
+For example, `RWBuffer<bool> b : register(u4);` will not emit any error in DXC, 
+but will in clang-dxc, despite the fact that `bool` is a valid RET.

```

Comment:
Per the link above, bool itself seems to work. Bool vectors do not work in any respect, but that's another matter: https://godbolt.org/z/eb7Yh73W6

Reply:
Hmmm, there must have been a fix over the past several months for there to no longer be any errors. I'll update this statment.